src/Avalonia.Controls/Grid/DefinitionBase.cs

```diff
@@ -0,0 +1,342 @@
+// Copyright (c) The Avalonia Project. All rights reserved.
+// Licensed under the MIT license. See licence.md file in the project root for full license information.
+
+using System;
+using System.Collections;
+using System.Diagnostics;
+
+namespace Avalonia.Controls
+{
+    /// <summary>
+    /// Base class for <see cref="ColumnDefinition"/> and <see cref="RowDefinition"/>.
+    /// </summary>
+    public abstract class DefinitionBase : AvaloniaObject
+    {
+        /// <summary>
+        /// Static ctor. Used for static registration of properties.
+        /// </summary>
+        static DefinitionBase()
+        {
+            SharedSizeGroupProperty.Changed.AddClassHandler<DefinitionBase>(OnSharedSizeGroupPropertyChanged);
+        }
+        
+        internal bool UseSharedMinimum { get; set; }
+        internal bool LayoutWasUpdated { get; set; }
+        
+        private int _parentIndex = -1;                       //  this instance's index in parent's children collection
+        private LayoutTimeSizeType _sizeType;      //  layout-time user size type. it may differ from _userSizeValueCache.UnitType when calculating "to-content"
+        private double _minSize;                        //  used during measure to accumulate size for "Auto" and "Star" DefinitionBase's
+        private double _measureSize;                    //  size, calculated to be the input contstraint size for Child.Measure
+        private double _sizeCache;                      //  cache used for various purposes (sorting, caching, etc) during calculations
+        private double _offset;                         //  offset of the DefinitionBase from left / top corner (assuming LTR case)
+        internal SharedSizeScope _privateSharedSizeScope;
+        private SharedSizeState _sharedState;           //  reference to shared state object this instance is registered with
+        private bool _successUpdateSharedScope;
+
+        /// <summary>
+        /// Defines the <see cref="SharedSizeGroup"/> property.
+        /// </summary>
+        public static readonly StyledProperty<string> SharedSizeGroupProperty =
+            AvaloniaProperty.Register<DefinitionBase, string>(nameof(SharedSizeGroup), inherits: true);
+
+        /// <summary>
+        /// Gets or sets the name of the shared size group of the column or row.
+        /// </summary>
+        public string SharedSizeGroup
+        {
+            get { return GetValue(SharedSizeGroupProperty); }
+            set { SetValue(SharedSizeGroupProperty, value); }
+        }
+        /// <summary>
+        /// Callback to notify about entering model tree.
+        /// </summary>
+        internal void OnEnterParentTree(Grid grid, int index)
+        {
+            Parent = grid;
+            _parentIndex = index;
+        }
+
+        internal void UpdateSharedScope()
```

Comment:
@grokys This function is being called all over the place, basically i need both PrivateSharedSizeScope and SharedSizeGroup not null when this runs so i resorted in placing calls on this method on both Measure/Arrange which is not ideal...